python_modules/dagster/dagster/_core/execution/context/compute.py

```diff
@@ -1223,8 +1245,467 @@ def asset_check_spec(self) -> AssetCheckSpec:
+
+    @property
+    def provenance(self) -> Optional[DataProvenance]:
+        return self.get_asset_provenance(self.asset_key)
+
+    @property
+    def provenance_by_asset_key(self) -> Mapping[AssetKey, Optional[DataProvenance]]:
+        provenance_map = {}
+        for key in self.asset_keys:
+            provenance_map[key] = self.get_asset_provenance(key)
+
+        return provenance_map
+
+    @property
+    def code_version(self) -> Optional[str]:
+        return self.get_assets_code_version([self.asset_key])[self.asset_key]
+
+    @property
+    def code_version_by_asset_key(self) -> Mapping[AssetKey, Optional[str]]:
+        return self.get_assets_code_version(self.asset_keys)
+
+    @public
+    @property
+    def is_partition_step(self) -> bool:
+        return self._op_execution_context.has_partition_key
+
+    @property
+    def partition_key(self) -> str:
+        return self.op_execution_context.partition_key
+
+    @public
+    @property
+    def partition_key_range(self) -> PartitionKeyRange:
+        return self._op_execution_context.asset_partition_key_range
+
+    @property
+    def partition_time_window(self) -> TimeWindow:
+        return self.op_execution_context.partition_time_window
+
+    @public
+    @property
+    def run_id(self) -> str:
+        return self._op_execution_context.run_id
+
+    @property
+    def job_name(self) -> Optional[str]:
+        return self.op_execution_context.job_name
+
+    @property
+    def retry_number(self) -> int:
+        return self.op_execution_context.retry_number
+
+    # Additional methods
+
+    @public
+    @property
+    def instance(self) -> DagsterInstance:
+        """DagsterInstance: The current Dagster instance."""
+        return self._step_execution_context.instance
+
+    @public
+    @property
+    def dagster_run(self) -> DagsterRun:
+        """PipelineRun: The current pipeline run."""
+        return self._step_execution_context.dagster_run
+
+    @public
+    @property
+    def pdb(self) -> ForkedPdb:
+        return self._op_execution_context.pdb
+
+    @public
+    @property
+    def log(self) -> DagsterLogManager:
+        """DagsterLogManager: The log manager available in the execution context."""
+        return self._op_execution_context.log
+
+    @public
+    def log_event(self, event: UserEvent) -> None:
+        return self._op_execution_context.log_event(event)
+
+    @public
+    @property
+    def assets_def(self) -> AssetsDefinition:
+        return self._op_execution_context.assets_def
+
+    @public
+    @property
+    def selected_asset_keys(self) -> AbstractSet[AssetKey]:
+        return self._op_execution_context.selected_asset_keys
+
+    @public
+    @experimental
+    def get_asset_provenance(self, asset_key: AssetKey) -> Optional[DataProvenance]:
+        return self._op_execution_context.get_asset_provenance(asset_key)
+
+    @public
+    # TODO - method naming. this needs work
```

Comment:
Alternate names:
* `get_code_versions_for_assets`


Alternately, we could delete this method entirely, and update the impls of `code_version` and `code_version_by_asset_key` to 
```python
    @property
    def code_version(self) -> Optional[str]:
        return self.code_version_by_asset_key[self.asset_key]

    @property
    def code_version_by_asset_key(self) -> Mapping[AssetKey, Optional[str]]:
        return self.op_execution_context.instance.get_latest_materialization_code_versions(
            self.asset_keys
        )
```

Reply:
It is important that access each in-memory properties and expensive accesses to the instance feel very different, and that it is clear when the user is doing something expensive like a db query to get information, so something like `code_version_by_asset_key` is not a good idea. Imagine accessing that in a tight loop.

Reply:
So to be equivalent with the ext contract (code here: https://github.com/dagster-io/dagster/blob/34acd9c10c1b938ab10f9e26042516e2dca71600/python_modules/dagster/dagster/_core/ext/context.py#L122) only code versions on the asset_defs object would be accessible cheaply. Do we need to have a the variant that fetches from the instance at all?

Reply:
`code_version`, `code_version_by_asset_key`, `provenance` and `provenace_by_asset_key` have all been updated so that the dictionaries are pre-fetched during context init. additionally `code_versions` are now pulled from the AssetsDefinition.

Reply:
>Does making it a getter method seem like enough indication to the user that it's expensive?

Maybe tangential to the immediate concern, but I like "fetch" as an indicator that a method is loading from elsewhere.

Reply:
ok great, i didn't know that subtlety (thinking about it now it makes total sense) so thanks for clearing that up